src/mlpack/methods/ann/augmented/tree_memory_impl.cpp

```diff
@@ -0,0 +1,93 @@
+/**
+ * @file tree_memory_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of CopyTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_TREE_MEMORY_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TREE_MEMORY_IMPL_HPP
+
+#include <cassert>
+
+#include "tree_memory.hpp"
+
+TreeMemory<T>::TreeMemory(size_t size, J joiner, W writer) {
+  memorySize = size;
+  // Rounding size to the next highest power of 2.
+  // WARNING: this is a hack that works only for 32-bit integers.
+  // In practice we don't really want to 
+  // allocate 1e12 memory cells anyway, though.
+  assert(0 < size && size < static_cast<size_t>(1 << 31));
+  actualMemorySize = size - 1;
+  actualMemorySize |= actualMemorySize >> 1;
+  actualMemorySize |= actualMemorySize >> 2;
+  actualMemorySize |= actualMemorySize >> 4;
+  actualMemorySize |= actualMemorySize >> 8;
+  actualMemorySize |= actualMemorySize >> 16;
+  actualMemorySize |= actualMemorySize >> 32;
+  actualMemorySize++;
```

Comment:
Honestly I think it would be easier and more portable to just do `ceil(log2((double) size))` here.